week04/792. Number of Matching Subsequences.cpp

```diff
@@ -0,0 +1,65 @@
+//792. Number of Matching Subsequences
+//S = total number of characters in words, N = length of string s
+//TC - O(s + Σ words)
+//SC - O( S)
+
+class Solution {
+public:
+    class TrieNode
+    {
+     public: 
+        int cnt = 0;
+        TrieNode* next[26];
+        TrieNode()
+        {
+            for(int i = 0; i < 26; i++)
+            {
+                next[i] = NULL;
+            }
+        }
+    };
+    TrieNode* generateTrieNode(vector<string>& words)
+    {
+        TrieNode* root = new TrieNode();
+        for(auto &wd : words)
+        {
+            TrieNode* currNode = root;
+            for(auto &c : wd)
+            {
+                int id = c - 'a';
+                if(currNode -> next[id] == NULL)
+                {
+                    currNode -> next[id] = new TrieNode();
+                }
+                currNode = currNode -> next[id];
+            }
+            currNode -> cnt++;
+        }
+        return root;
+    }
+    
+    
+    int dfs(string& s, TrieNode* u, int pos)
+    {
+        if(u == NULL) return 0;
+        
+        int ans = u -> cnt;
+        if (pos >= s.length())return ans;
+     
+        for(int i = 0; i < 26; i++)
+        {
+            if (u->next[i])
+            {
+                int index = s.find('a' + i, pos);
```

Comment:
This is linear in time complexity. So the overall worst-case time complexity is `O(S * Σ words)`. Though it barely reaches that, it is the one.